src/corelibs/U2View/src/ov_sequence/view_rendering/SequenceViewAnnotatedRenderer.cpp

```diff
@@ -188,16 +187,56 @@ void SequenceViewAnnotatedRenderer::drawAnnotation(QPainter& p, const QSize& can
                     const QString aText = prepareAnnotationText(aData, as);
                     drawBoundedText(p, annotationRect, aText);
                 }
-                if (simple &&
-                    annotationRect.width() > MIN_WIDTH_TO_DRAW_EXTRA_FEATURES &&
-                    displaySettings.displayCutSites) {
-                    drawCutSite(p, aData, r, annotationRect, as->color, canvasSize, visibleRange);
-                }
             }
             drawAnnotationConnections(p, a, as, displaySettings, canvasSize, visibleRange);
-        } else if (isRestrictionSite && simple && displaySettings.displayCutSites) {
-            p.setPen(borderPen);
-            drawCutSite(p, aData, r, QRect(), as->color, canvasSize, visibleRange);
+        }
+    }
+    // draw restriction sites
+    /**
+     * Calculates restriction cuts positions depends on visible area and calls drawCutSite.
+     *
+     * \param reg Region of the annotation, which sites should be drawn.
+     * \param drawDirect Draw direct cut if true and do not draw if false.
+     * \param drawComplement Draw complement cut if true do not draw if false.
+     */
+    auto drawCutSiteLambda = [this, &p, &visibleRange, &a, &as, &availableHeight, &canvasSize, selected, &aData](const U2Region& reg, bool drawDirect = true, bool drawComplement = true) {
+        if (reg.intersects(visibleRange)) {
+
+            const U2Region visibleLocation = reg.intersect(visibleRange);
+            const U2Region y = getAnnotationYRange(a, 0, as, availableHeight);
+            CHECK(y.startPos >= 0, );
```

Comment:
cut-site may be on the different y-level from the annotation region (outside of the annotation region). See AbaAI enzymes.

Reply:
You are right: it can be a bug in the old code: in this case the bug should be addressed in a new PR. I not ask you to fix in this PR. But if you add a new code with a copied bug this is not good. The new code should not have the bug.

The biggest problem in your patch is the following: 
1) You separated cut-site drawing into a dedicated block: this is very good.
2) You rely on annotation regions for cut-sites geometry calculation, while cut-sites may be out of annotation regions: this is incorrect and should be fixed.

Look these examples for cut-site drawing related to the wrong geometry:

1) Good: Annotation fits X region.
![image](https://github.com/ugeneunipro/ugene/assets/17791039/0bce5eef-83e7-491c-bbdb-4544afb00e61)

2) Bad: Annotation does not fit X region: direct cut-site is not drawn.
![image](https://github.com/ugeneunipro/ugene/assets/17791039/e923cf45-715b-4388-b4e3-8ea0d4f76337)

For the wrap-mode we have an issue with Y:
1) Good, no wrap inside cut-site region
![image](https://github.com/ugeneunipro/ugene/assets/17791039/07d38f85-60fe-4eae-a61d-788bd2689d7a)

2) Bad: cutsite has wrong Y position
![image](https://github.com/ugeneunipro/ugene/assets/17791039/95e1af25-1b21-43fa-a7f1-fd539fa3f958)

A good API test can be created if you extract a function that does not draw anything, but only calculate a cut-site layout in sequence (annotation regions) coordinates.

The function should accept as parameters: 
1) One annotation region (enzyme region)
2) a cutsite offset info

and will return another annotation region of length 1: where to draw the cut-site.

Next you can use 'annotation-region' logic to map the cutsite into screen and decide of to draw it (if it is visible) or not (if it out of the screen)

Reply:
OK, I got the point, just have some question, to be certain:

1. What do you mean by "a cutsite offset info". Now, calculation of cut regions required very many different parameters (`canvasSize`, `visibleRange`, etc). Do you mean here "anything you need, but try at least to minimize the number of arguments"?
2. "will return another annotation region of length 1: where to draw the cut-site." - this is a pretty simple function, are you sure, that you want to extract to the separate function this exact part? In this case, we will need to extract [this part](https://github.com/ugeneunipro/ugene/pull/1338/files#diff-f378eee993cc84932724155c8f4dace9314d18d12c76c43d909129c04b233accR334-R359) of code to separate function - it would be enough to know sequence region of cut. But, even when we know this region, we still do need to process some calculations, to find out screen positions where we need to draw these cites (if we scroll sequence cut region remains the same, but screen position become different). Maybe, it would be better to calculate [CutSiteDrawData](https://github.com/ugeneunipro/ugene/blob/815c4f31ad3b2b853eec59bc7a01ee2024b6d9a9/src/corelibs/U2View/src/ov_sequence/view_rendering/SequenceViewAnnotatedRenderer.h#L46), which have full information about "where this cut site should be drawn"?

Reply:
> What do you mean by "a cutsite offset info". 

I suggest to make this function screen-layout independent: return in-sequence position of the cut-site (like an annotation region). 
This way you can test this function using API tests without any screen geometry.

>"will return another annotation region of length 1: where to draw the cut-site." - this is a pretty simple function, are you sure, that you want to extract to the separate function this exact part?

Yes, simple functions with a small amount of parameters are always good.  They are readable and testable.

What we have in the current solution is a lambda that accepts a lot of parameters and depends on the current context and you said you can't even test it.

As the result there may be these functions created
1) Given an annotation calculate cut-site coordinates in sequence coordinates
2) Given a cut-site  sequence coordinates calculate on-screen coordinates

Reply:
OK, I tried to do something, just check